easybuild/framework/easyconfig/tweak.py

```diff
@@ -795,7 +802,107 @@ def map_toolchain_hierarchies(source_toolchain, target_toolchain, modtool):
+
+    :param software_name: Name of software
+    :param software_version_stub: initial characters of version
+    :param original_toolchain: original toolchain
+    :param toolchain_mapping: toolchain mapping from that containing original to target
+    :return: dictionary of possible mappings
+    """
+    orig_toolchain_hierarchy = get_toolchain_hierarchy(original_toolchain)
+    target_toolchain_hierarchy = get_toolchain_hierarchy(toolchain_mapping[original_toolchain['name']])
+
+    versionsuffix_mappings = {}
+
+    # Find highest value in the target
+    target_version = '0'
+    for toolchain in target_toolchain_hierarchy:
+        if toolchain['name'] == DUMMY_TOOLCHAIN_NAME:
+            toolchain_suffix = EB_FORMAT_EXTENSION
+        else:
+            toolchain_suffix = "-%s-%s" % (toolchain['name'], toolchain['version'])
+        regex_search_query = '^%s-%s.*' % (software_name, software_version_stub) + toolchain_suffix
+        versionprefix_name = '%s-' % software_name
+
+        cand_paths = search_easyconfigs(regex_search_query, return_robot_list=True)
+        for path in cand_paths:
+            # Get the version from the path
+            filename = os.path.basename(path)
+            # Find the version sandwiched between our known values
+            try:
+                regex = '%s(.*)%s' % (versionprefix_name, toolchain_suffix)
+                version = re.search(regex, filename).group(1)
+                if LooseVersion(version) > LooseVersion(target_version):
+                    target_version = version
+            except AttributeError:
+                raise EasyBuildError("Somethings wrong, could not extract version from %s using %s", filename,
+                                     regex)
+
+    # Now map all matching versions in the source toolchain to this target
+    if target_version > 0:
+        source_versions = set()
+        for toolchain in orig_toolchain_hierarchy:
+            if toolchain['name'] == DUMMY_TOOLCHAIN_NAME:
+                toolchain_suffix = EB_FORMAT_EXTENSION
+            else:
+                toolchain_suffix = "-%s-%s" % (toolchain['name'], toolchain['version'])
+            regex_search_query = '^%s-%s.*' % (software_name, software_version_stub) + toolchain_suffix
+            versionprefix_name = '%s-' % software_name
+
+            cand_paths = search_easyconfigs(regex_search_query, return_robot_list=True)
+            for path in cand_paths:
+                # Get the version from the path
+                filename = os.path.basename(path)
+                # Find the version sandwiched between our known values
+                try:
+                    regex = '%s(.*)%s' % (versionprefix_name, toolchain_suffix)
+                    version = re.search(regex, filename).group(1)
+                    if LooseVersion(version) < LooseVersion(target_version):
+                        source_versions.add(version)
```

Comment:
A *very* large part of this is duplicate code, this should definitely go into a (sub-) function rather than copy-pasting it...